### PR TITLE
SONARJAVA-1172 Looks for PreparedStatement last assignement

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR Add a 'package-info.java' file to document the 'src/main/java/org/sonar/java/checks' package (org.sonarsource.java:java-checks:src/main/java/org/sonar/java/checks) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1228)

File: java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java [hunk 1]
```diff
@@ -120,4 +171,35 @@ private static Integer getNumberParametersFromPreparedStatement(@Nullable Tree t
...
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (isSearchedVariable(tree.variable())) {
+        reassignments.add(tree);
+      }
+      super.visitAssignmentExpression(tree);
+    }
+
+    private boolean isSearchedVariable(ExpressionTree variable) {
+      return variable.is(Tree.Kind.IDENTIFIER) && usages.contains((IdentifierTree) variable);
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java]: @m-g-sonar `(IdentifierTree)` cast not needed.

File: java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java [hunk 2]
```diff
@@ -120,4 +171,35 @@ private static Integer getNumberParametersFromPreparedStatement(@Nullable Tree t
...
   }
+
+  private static class ReassignmentFinder extends BaseTreeVisitor {
+
+    private final List<IdentifierTree> usages;
+    private List<Tree> reassignments;
+
+    public ReassignmentFinder(List<IdentifierTree> usages) {
+      this.usages = usages;
+      this.reassignments = Lists.newLinkedList();
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java]: @m-g-sonar you can use new LinkedList<>() thanks to java 7.

File: java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java [hunk 3]
```diff
@@ -95,23 +103,66 @@ protected void onMethodInvocationFound(MethodInvocationTree mit) {
...
+    return result;
+  }
+
+  private List<Tree> getReassignments(Tree ownerDeclaration, List<IdentifierTree> usages) {
+    if (ownerDeclaration != null) {
+      ReassignmentFinder reassignmentFinder = new ReassignmentFinder(usages);
+      ownerDeclaration.accept(reassignmentFinder);
+      return reassignmentFinder.reassignments;
+    }
+    return Lists.newArrayList();
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java]: @m-g-sonar you can use new ArrayList<>() thanks to java 7.

File: java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java [hunk 4]
```diff
@@ -95,23 +103,66 @@ protected void onMethodInvocationFound(MethodInvocationTree mit) {
...
+    if (lastReassignment != null) {
+      return lastReassignment;
+    }
+    return result;
+  }
+
+  @CheckForNull
+  private Tree getLastReassignment(int line, Symbol referenceSymbol) {
+    Tree result = null;
+    for (Tree reassignement : reassignmentBySymbol.get(referenceSymbol)) {
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java]: @m-g-sonar reassignement -> reassignment

File: java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java [hunk 5]
```diff
@@ -95,23 +103,66 @@ protected void onMethodInvocationFound(MethodInvocationTree mit) {
...
+      int reassignementLine = FirstSyntaxTokenFinder.firstSyntaxToken(reassignement).line();
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java]: @m-g-sonar reassignementLine -> reassignmentLine

File: java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java [hunk 6]
```diff
@@ -86,7 +98,7 @@ protected void onMethodInvocationFound(MethodInvocationTree mit) {
         addIssue(mit, "PreparedStatement indices start at 1.");
       } else {
         Tree preparedStatementReference = getPreparedStatementReference(mit);
-        Integer numberParameters = getNumberParametersFromPreparedStatement(preparedStatementReference);
+        Integer numberParameters = getPreparedStatementNumberOfParameters(preparedStatementReference);
         if (numberParameters != null && methodFirstArgumentValue > numberParameters.intValue()) {
```
Comment (mpaladin) [inline java-checks/src/main/java/org/sonar/java/checks/PreparedStatementAndResultSetCheck.java]: @m-g-sonar unnecessary unboxing `.intValue()`. Doesn't eclipse notify you about it?

**Pull request merged**